caas/kubernetes/provider/resources/applier.go

```diff
@@ -52,29 +56,73 @@ func (op *operation) process(ctx context.Context, rollback Applier) error {
 	} else if err != nil {
 		return errors.Annotatef(err, "checking if resource %q exists or not", existingRes.ID())
 	}
-	if found {
-		ver := op.resource.GetObjectMeta().GetResourceVersion()
-		if ver != "" && ver != existingRes.GetObjectMeta().GetResourceVersion() {
-			id := op.resource.ID()
-			return errors.Annotatef(errConflict, "%s %s", id.Type, id.Name)
-		}
-	}
+
 	switch op.opType {
 	case opApply:
-		err = op.resource.Apply(ctx)
-		if found {
-			// apply the previously existing resource.
-			rollback.Apply(existingRes)
-		} else {
-			// delete the new resource just created.
-			rollback.Delete(op.resource)
+		err = retry.Call(retry.CallArgs{
+			Func: func() error {
+				err := op.resource.Apply(ctx)
+				if errors.Is(err, errConflict) {
+					_ = op.resource.Get(ctx) // refresh resource version
+					return err
+				}
+				// this might happen if it were deleted by another apiserver.
+				if k8serrors.IsNotFound(err) {
+					return err
+				}
+				return errors.Annotatef(err, "applying resource %q", op.resource.ID().Name)
+			},
+			IsFatalError: func(err error) bool {
+				return !k8serrors.IsConflict(err) && !errors.Is(err, errConflict)
+			},
+			Clock:       jujuclock.WallClock,
+			Attempts:    5,
+			Delay:       time.Second,
+			BackoffFunc: retry.ExpBackoff(time.Second, 5*time.Second, 1.5, true),
+		})
+
+		// we do not want to rollback if the resource is not found, not applied or
+		// it has been deleted by another apiserver during the retry call process.
+		if err == nil {
+			if found {
+				// apply the previously existing resource.
+				rollback.Apply(existingRes)
+			} else {
+				// delete the new resource just created.
+				rollback.Delete(op.resource)
+			}
 		}
 	case opDelete:
-		err = op.resource.Delete(ctx)
-		if found {
+		// delete with retry
+		err = retry.Call(retry.CallArgs{
+			Func: func() error {
+				err := op.resource.Delete(ctx)
+				if k8serrors.IsConflict(err) {
```

Comment:
Can we get a conflict when deleting? It would only be a not found right?

Reply:
I remembered making this retry change for delete because I traced the bug an err in the custom resource delete method, thought it was due to lightkube del conflict. Just did a check and it seems fine now, will remove retry for delete. Thanks ian!

https://github.com/juju/juju/pull/20489/commits/67797fadfbc27196d1ec561aa80a4008fe8a2ae6